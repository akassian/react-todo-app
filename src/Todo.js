import React, { useState } from "react";
import UpdateTodoForm from "./UpdateTodoForm";

const deleteLabel = "X";

function Todo({ id, message, remove, toggleIsEdit, updateTodo, isEdit }) {
  const [isDone, setIsDone] = useState(false);
  return (
    <li data-testid={`${message}`}>
      {isDone ? (
        <p>
          <s>{message}</s>
        </p>
      ) : (
        <p>{message}</p>
      )}

      <button
        data-testid={`${message}Remove`}
        style={{ marginLeft: 10 }}
        onClick={(evt) => remove(id)}
      >
        {deleteLabel}
      </button>
      <button
        data-testid={`${message}Edit`}
        style={{ marginLeft: 10 }}
        onClick={(evt) => toggleIsEdit()}
      >
        Edit
      </button>
      {isEdit ? (
        <UpdateTodoForm updateTodo={updateTodo} message={message} id={id} />
      ) : null}
      <input
        id="done"
        type="checkbox"
        checked={isDone}
        onChange={(evt) => setIsDone(!isDone)}
      />
    </li>
  );
}

export default Todo;
export { deleteLabel };
