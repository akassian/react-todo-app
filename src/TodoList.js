import React, { useState } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import uuid from "uuid/v4";

function TodoList() {
  const [todos, setTodos] = useState([]);
  const [isEdit, setIsEdit] = useState(false);

  const renderedTodoList = () => {
    return (
      <ol>
        {todos.map((todo) => (
          <Todo
            key={todo.id}
            id={todo.id}
            message={todo.message}
            remove={remove}
            updateTodo={updateTodo}
            toggleIsEdit={toggleIsEdit}
            isEdit={isEdit}
          />
        ))}
      </ol>
    );
  };

  /** Add a new todo. */

  const addTodo = (todo) => {
    let newTodo = { ...todo, id: uuid() };
    setTodos((todos) => [...todos, newTodo]);
  };

  /** Remove a todo. */

  const remove = (id) => {
    setTodos((todos) => todos.filter((todo) => todo.id !== id));
  };

  // /** Update a todo. */

  const updateTodo = (id, message) => {
    setTodos(
      todos.map((todo) => (todo.id === id ? { ...todo, message } : todo))
    );
  };

  const toggleIsEdit = () => {
    setIsEdit(!isEdit);
  };

  return (
    <div className="TodoList">
      {isEdit ? <h1>Edit Todo List!</h1> : <h1>Add to Todo List!</h1>}
      <NewTodoForm addTodo={addTodo} />
      {renderedTodoList()}
    </div>
  );
}

export default TodoList;
