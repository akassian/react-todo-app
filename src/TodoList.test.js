import React from "react";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import TodoList from "./TodoList";
import { deleteLabel } from "./Todo";

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a new todo", function () {
  const { getByLabelText, queryByText, queryByTestId } = render(<TodoList />);

  // no remove button or a todo yet
  expect(queryByText(deleteLabel)).not.toBeInTheDocument();

  const message = getByLabelText("Message:");
  const submitBtn = queryByText("Add a Todo!");

  // fill out the form
  fireEvent.change(message, { target: { value: "first" } });
  fireEvent.click(submitBtn);

  // remove button and a todo exists!
  expect(queryByTestId("first")).toBeInTheDocument();
  expect(queryByTestId("firstRemove")).toBeInTheDocument();
});

it("can remove a corressponding todo", function () {
  const { getByLabelText, queryByText, queryByTestId } = render(<TodoList />);

  // no remove button yet or todoes
  expect(queryByText(deleteLabel)).not.toBeInTheDocument();

  const message = getByLabelText("Message:");
  const submitBtn = queryByText("Add a Todo!");

  // Make two todos
  fireEvent.change(message, { target: { value: "first" } });
  fireEvent.click(submitBtn);

  fireEvent.change(message, { target: { value: "second" } });
  fireEvent.click(submitBtn);

  //remove red todo
  const removeBtn = queryByTestId("firstRemove");
  fireEvent.click(removeBtn);

  //red todo disappears and black todo stays
  expect(queryByTestId("first")).toBeNull();
  expect(queryByTestId("firstRemove")).toBeNull();
  expect(queryByTestId("second")).toBeInTheDocument();
  expect(queryByTestId("secondRemove")).toBeInTheDocument();
});
