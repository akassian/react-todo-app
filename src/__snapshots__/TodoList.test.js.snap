// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <h1>
      Add to Todo List!
    </h1>
    <form>
      <div>
        <label
          for="message"
        >
          Message:
        </label>
        <input
          id="message"
          name="message"
          value=""
        />
      </div>
      <br />
      <button>
        Add a Todo!
      </button>
    </form>
    <ol />
  </div>
</DocumentFragment>
`;
